File: Backtracking/22.Generate Parentheses/Solution.py
class Solution:
    # @param {integer} n
    # @return {string[]}
	'''
		找出给定数目的有效括号组合（不能出现”)("
		属于搜索问题，可用DFS策略的回溯搜索
		参数含义：left代表剩余左括号数，right表示剩余右括号数，str代表当前临时解
	'''
    def dfs(self, left, right, result, str):
        if left == 0 and right ==0 :
            result.append(str)   # 左右括号都用完了，递归终止。
            return
        if left > 0:
			#每个左括号，添加一个'('到临时解，左括号数减一，对应右括号数加一(即对应的右括号)
            self.dfs(left-1, right+1, result, str+'(') 
        if right > 0:
			#每个右括号，添加一个’)'到临时解，只需把右括号数减一即可
            self.dfs(left, right-1, result, str+')') #
    
    def generateParenthesis(self, n):
        result = []
        self.dfs(n, 0, result, '')  #n对括号，初始n个左括号，递归过程中会添加对应的右括号
        return result

File: Backtracking/22.Generate Parentheses/Solution.java
/**
 * 给定n，输出所有合法匹配的左右括号组合
 * 题目地址：https://leetcode.com/problems/generate-parentheses/
 * 类似题目：Letter Combinations of a Phone Number， Valid Parentheses
**/
public class Solution {

    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<> ();
        dfs(result, n, n, "");
        return result;
    }
    
	/**
	 * dfs回溯搜索的另一种写法
     * left代表左括号数，right代表右括号数，answer代表有效答案
	**/
    private void dfs(List<String> result, int left, int right, String answer) {
        if( left > right) return;   //避免当前左括号数大于右括号数，这样会出现左括号对应右括号)(的逆序
        if( left == 0 && right == 0) {
            result.add(answer);
            return;
        }
		if( left > 0) dfs(result, left-1, right, answer+"(");  //每次处理左括号时，不添加对应右括号数，则需要添加回溯判断条件
        if( right > 0) dfs(result, left, right-1, answer+")"); 
    }
    
    
	/**
	 *	一种动态规划的搜索写法，迭代计算, 自底向上。
	 *	设f(n)为对应n对括号的解，依次计算f(0)...f(i)...f(n)，f(n)可以拆分为其它f(i)的组合
	 *	拆分的递推公式为f(i): (f(j))f(i-j-1)。其中j = 0...j-1
	**/
    public List<String> generateParenthesis2(int n) {
        List<String>[] resultArray = new List[n+1];  //保存f(i)的list数组，0<=i<=n
        resultArray[0] = Arrays.asList("");
        for(int i = 1; i <=n; i++) {  //迭代计算f(1)...f(n)的解
            List<String> tmp = new ArrayList<> ();  
            for(int j = 0; j < i; j++) {   //计算f(i)的解
                for( String s1 : resultArray[j]) {
                    for( String s2 : resultArray[i-j-1]) {
                        tmp.add( "(" + s1 + ")" + s2);
                    }
                }
            }
            resultArray[i] = tmp;
        }
        return resultArray[n];
    }
}